feat: finish up UI changes and code styling

BZahory opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/PR?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Commit:
chore: create util for swap+send quotes

Commit:
fix: clear quote data when switching to send

Commit:
chore: prevent scroll wiggle

update flexbox stylings

Commit:
chore: merge files after rebase

Commit:
fix: only reset gas limit if no swap

Commit:
fix: prevent infinite useeffect loop

aggregate state vars

Commit:
chore: remove no conversion rate text

Commit:
chore: prevent special symbols in input

Commit:
chore: tweak icon size

Commit:
feat: preserve swap+send form on edit

Commit:
fix: update variable name

broken from rebase

Commit:
fix: reset amount when sender changes

Commit:
fix: prevent console error for undefined

Commit:
chore: remove p > div nesting

removes a console error

Commit:
chore: use statically checkable deps

Commit:
docs: update task list

Commit:
chore: only show TOS when quote is fetched

Commit:
perf: reuse formatted value

Commit:
perf: reorder dispatches

Commit:
chore: avoid deconstruction

Commit:
perf: reduce action holdup

Commit:
perf: debounce computed gas limit

Commit:
perf: clean up timeout

Commit:
chore: use useCallback

Commit:
refactor: avoid usage of `useTokensToSeach`

micaelae committed:
[23849] fix: negative max token amount (#24447)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24447?quickstart=1)

## **Related issues**

Fixes:
[23849](https://github.com/MetaMask/metamask-extension/issues/23849)

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -2054,7 +2073,22 @@ export function editExistingTransaction(assetType, transactionId) {
-    if (assetType === AssetType.native) {
+    const isSwapAndSend = state[name].prevSwapAndSendInput;
+
+    if (isSwapAndSend) {
+      dispatch(
+        actions.addNewDraft({
+          ...draftTransactionInitialState,
+          ...state[name].prevSwapAndSendInput,
+          id: transactionId,
+          fromAccount: account,
+          history: [
+            `sendFlow - user clicked edit on transaction with id ${transactionId} (swap and send)`,
+          ],
+        }),
+      );
+      dispatch(updateSendQuote());
+    } else if (assetType === AssetType.native) {
```
Did you consider decoding the txData instead to extract the draft transaction details? The decoder can also be used to populate the recipient in the confirmation page (we currently show the swaps contract address)

BZahory replied:
The swaps-recipient bug was a different that's since been resolved. However, you're 100% right about decoding being preferred. I realize that we lose the `prevSwapAndSendInput` whenever the extension is closed so that edge makes this worth implementing.

BZahory replied:
Side note but it always amazes me how well you can pick up on what's going on in code – no doubt a huge contributor to the quality of your reviews. You're by far the best reviewer I've come across and I really do appreciate that ❤️

Commit:
fix: swap+send quote restoration

Commit:
refactor: move state update to cases

Commit:
fix: use max value instead of zero

Commit:
fix: ensure amount is populated correctly

Commit:
fix: ensure hooks are called consistently

Commit:
chore: update font weight

Commit:
feat: show 0 nft error after first touch

Commit:
fix: restore undefined onAmountChange behavior

Commit:
chore: ensure top token sorting

Commit:
fix: allow for multiple repeats

Commit:
chore: enable max when fiat is primary

Commit:
fix: update selector usage after rebase

Commit:
refactor: simplify return

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -2188,24 +2245,36 @@ export function updateSendQuote(
       draftTransaction?.sendAsset?.details?.address !==
       draftTransaction?.receiveAsset?.details?.address;
 
+    const {
+      quotes,
+      swapQuotesError,
+      isSwapQuoteLoading,
+      swapQuotesLatestRequestTimestamp,
+    } = draftTransaction;
```
I think we need to either return early if there's no draft transaction, or add a fallback if draft tx is undefined like this: 

```suggestion
    const {
      quotes,
      swapQuotesError,
      isSwapQuoteLoading,
      swapQuotesLatestRequestTimestamp,
    } = draftTransaction ?? {};
```

BZahory replied:
There should be a `draftTransaction` anytime the user is one the send flow (see: `initializeSendState`):

```
    if (!draftTransaction) {
      return thunkApi.rejectWithValue(
        'draftTransaction not found, possibly not on send flow',
      );
    }

BZahory replied:
Doesn't hurt to have it though so we can add the object fallback.

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-modal/asset-picker-modal.tsx:
```diff
@@ -111,6 +123,26 @@ export function AssetPickerModal({
+    numberOfDecimals: primaryNumberOfDecimals,
+  } = useUserPreferencedCurrency(PRIMARY, { ethNumberOfDecimals: 4 });
+
+  const [, primaryCurrencyProperties] = useCurrencyDisplay(balanceValue, {
+    numberOfDecimals: primaryNumberOfDecimals,
+    currency: primaryCurrency,
+  });
```
Is there another util we can use to get this value? When the user's balance is small, `primaryCurrencyProperties.value` is a human-readable string like `< 0.000001`, which causes `getTokenFiatAmount` to throw an error

micaelae replied:
Uploading Screen Recording 2024-05-10 at 2.36.12 PM.mov…

micaelae replied:
https://github.com/MetaMask/metamask-extension/assets/100321200/a45c4f1f-7c0b-4260-89ce-3ef6249fc956

BZahory replied:
This error was related to an issue in a pair of utils we depend on; updated in 0965dd66a42a0ebef8829f7d0fe2e82f8a146408

micaelae commented on ui/components/multichain/pages/send/components/quote-card/hooks/useEthFeeData.tsx:
```diff
@@ -31,12 +31,15 @@ export default function useEthFeeData(gasLimit = 0) {
+  const { medium, gasPrice: maybeGasFee } = useSelector(getGasFeeEstimates);
+
+  const gasFee1559 = maybeGasFee ?? medium;
 
```
Seeing a bug where `ethGasFee` is calculated below. If networkAndAccountSupports1559=false and maybeGasFee is undefined, `medium` is converted into a BigNumber, which throws an exception.

This fixes it locally for me
```suggestion
  const gasFee1559 = maybeGasFee ?? medium?.suggestedMaxFeePerGas;
```

micaelae replied:
![Screenshot 2024-05-10 at 2 52 40 PM](https://github.com/MetaMask/metamask-extension/assets/100321200/d71ca12d-7e72-44d7-905e-40a4caabf8e1)

BZahory replied:
Updated

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -2834,6 +2850,7 @@ export function signTransaction() {
           sendAsset: { ...sendAsset },
           receiveAsset: { ...receiveAsset },
           recipient: { ...recipient },
+          amountMode: state[name].amountMode,
```
Not related to this commit - can we disable the Confirm button when Swap + send-ing with 0 amount? It's supported for simple Send (useful for cancelling txs, etc), but I don't see a usecase for using a 0-amount swap trade

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-modal/asset-picker-modal.tsx:
```diff
@@ -179,12 +179,19 @@ export function AssetPickerModal({
+
+      // topTokens should already be sorted by popularity
+      for (const address of Object.keys(topTokens)) {
+        const token = tokenList?.[address];
```
Do the cases of topTokens and tokenList addresses always match?

BZahory replied:
`topTokens` doesn't map to tokens but rather to the position of the token (e.g., `{'index': 2}`)

BZahory replied:
<img width="349" alt="Screenshot 2024-05-14 at 2 50 22 PM" src="https://github.com/MetaMask/metamask-extension/assets/44588480/19edaf0e-06a9-4e32-99c7-66b4beaa7f4f">

micaelae replied:
I meant the cases of their keys. Looks like they both use lowercase addresses

micaelae commented on ui/components/multichain/asset-picker-amount/swappable-currency-input/swappable-currency-input.tsx:
```diff
@@ -93,6 +98,7 @@ export function SwappableCurrencyInput({
       )}
       asset={asset?.details}
       isSkeleton={isAmountLoading}
+      isMatchingUpstream={isSetToMax}
```
Why not just call `const isSetToMax = useSelector(getSendMaxModeState);` in CurrencyInput?

BZahory replied:
`CurrencyInput` should be decoupled from Send logic (see: https://github.com/MetaMask/metamask-extension/pull/23192#discussion_r1526686978)

micaelae replied:
Gotcha, I think the naming could be improved then - maybe `isValueModifiedUpstream`

BZahory replied:
Isn't starting with "is" convention?

micaelae replied:
hmm good point - looks like is/has/can is the convention. I'm good with any prefix as long as the name is more descriptive

BZahory replied:
Hmm yeah, it's tricky because it's technically always syncing with the upstream; this explicitly says that it will always *match* the upstream

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx:
```diff
@@ -62,14 +65,27 @@ export const AssetPickerAmount = ({
+
+  const handleChange = useCallback(
+    (newAmountRaw, newAmountFormatted) => {
+      if (!isNFTInputChanged && asset.type === AssetType.NFT) {
+        setIsTokenInputChanged(true);
+      }
```
Do you think this state should be reset when the selected NFT token changes?

BZahory replied:
Yeah good call – updated

micaelae commented on ui/components/app/currency-input/hooks/useStateWithFirstTouch.tsx:
```diff
@@ -0,0 +1,16 @@
+// symbol to cover edge case where initial value is literally 'INITIAL_VALUE'
+const INITIAL_VALUE = Symbol('INITIAL_VALUE');
+
+export default function useStateWithFirstTouch<T>(
```
Which bug does this address?

BZahory replied:
It lets us know if the token amount has been modified by a user input in `CurrencyInput` so we can know if the amount is being updated upstream after opening the send flow via the "EDIT" flow

micaelae commented on ui/components/app/currency-input/currency-input.js:
```diff
@@ -178,11 +196,7 @@ export default function CurrencyInput({
-          {t('noConversionRateAvailable')}
-        </div>
-      );
+      return null;
```
Not related to this line - can we disable the Confirm button when Swap + send-ing with 0 amount? It's supported for simple Send (useful for cancelling txs, etc), but I don't see a usecase for using a 0-amount swap trade

micaelae replied:
Still seeing it locally. I'm able to click on Confirm, but it doesn't do anything. Doesn't attempt to fetch quotes either

![Screenshot 2024-05-14 at 4 44 32 PM](https://github.com/MetaMask/metamask-extension/assets/100321200/9b484aed-ac62-4112-86ca-323967b3dff1)

BZahory replied:
Oh my bad I thought you meant the error state – generalized now: b4ebf6e41dab6861384c0af82db6ab6a03953a58

Commit:
fix: update reference to property

Commit:
chore: add validation case

BZahory committed:
chore: add fallback

Co-authored-by: micaelae <100321200+micaelae@users.noreply.github.com>

Commit:
chore: reset state variable when asset changes

Commit:
fix: handle special string case

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx:
```diff
@@ -55,21 +57,36 @@ export const AssetPickerAmount = ({
+
+  const handleChange = useCallback(
+    (newAmountRaw, newAmountFormatted) => {
+      if (!isNFTInputChanged && asset.type === AssetType.NFT) {
```
How come the same error handling is not done for tokens?

BZahory replied:
We allow for 0 amount ERC20 transfers but not NFTs – not sure though

micaelae commented on ui/helpers/utils/token-util.js:
```diff
@@ -226,7 +226,12 @@ export function getTokenFiatAmount(
 
-  let currentTokenInFiat = new Numeric(tokenAmount, 10);
+  // the `string` property of a token is typically passed as `tokenAmount`; for very low values, it will be "< 0.000001"
+  const isTokenAmountNearZero = tokenAmount.includes('<');
```
`tokenAmount` can be a string or a number; this throws an error if it's the latter. Moreover, adding this edge case here seems like a fragile way to handle small token amounts

I'd recommend removing `string: primaryCurrencyProperties.value` in asset-picker-modal, to avoid this edge case here and in useTokensToSearch. Doing so causes the fallback string representation to show up, which seems acceptable
![Screenshot 2024-05-14 at 3 54 36 PM](https://github.com/MetaMask/metamask-extension/assets/100321200/c824aadb-0839-4b40-ad18-8f1ff84234b8)

BZahory replied:
I think this had to do with how I was reproing this issue. Mind checking if you see it on the latest branch?

micaelae replied:
LGTM

micaelae commented on ui/components/multichain/asset-picker-amount/nft-input/nft-input.tsx:
```diff
@@ -22,7 +22,7 @@ export function NFTInput({ integerValue, onChange, className }: NFTInputProps) {
+    onChange(newValue.toPrefixedHexString(), String(newValueAsString));
   };
 
   return (
```
Is there a way to make the amount input autofocus? That's the behavior for tokens, but doesn't seem to work for NFTs

BZahory replied:
`autofocus` is super inconsistent so I just added logic to focus on key press if that's okay

micaelae replied:
I see it working when I click Send from the NFTs tab in the overview. But not if I go to the NFTs tab from within Send. Low priority tho, so the latest version lgtm

BZahory replied:
Hmm I see it working for both

Commit:
fix: cast to string before calling method

Commit:
fix: generalize swap and send state validation

Commit:
refactor: move fix upstream

Commit:
chore: fully revert changes

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-modal/asset-picker-modal.tsx:
```diff
@@ -111,6 +123,26 @@ export function AssetPickerModal({
+    numberOfDecimals: primaryNumberOfDecimals,
+  } = useUserPreferencedCurrency(PRIMARY, { ethNumberOfDecimals: 4 });
+
+  const [, primaryCurrencyProperties] = useCurrencyDisplay(balanceValue, {
+    numberOfDecimals: primaryNumberOfDecimals,
+    currency: primaryCurrency,
+  });
+
```
This can be removed now
```suggestion

```

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-modal/asset-picker-modal.tsx:
```diff
@@ -123,30 +155,92 @@ export function AssetPickerModal({
+    nativeCurrency,
+    nativeCurrencyImage,
+    balanceValue,
+    primaryCurrencyProperties.value,
```
This can also be removed
```suggestion
```

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-modal/asset-picker-modal.tsx:
```diff
@@ -38,16 +37,27 @@ import { AssetType } from '../../../../../shared/constants/transaction';
 import { useEqualityCheck } from '../../../../hooks/useEqualityCheck';
-import { useTokensToSearch } from '../../../../hooks/useTokensToSearch';
-import { TokenBucketPriority } from '../../../../../shared/constants/swaps';
+import { useCurrencyDisplay } from '../../../../hooks/useCurrencyDisplay';
```
Can be removed
```suggestion
```

Commit:
feat: focus input on key press

Commit:
chore: remove unused variables

ejwessel committed:
Fix Issue#24491 (#24535)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24535?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24491

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Signed-off-by: Ethan Wessel <ethan.wessel@consensys.net>

Commit:
style: align to lint and import rules

micaelae committed:
[METABRIDGE-843] fix: compare lowercase recipients when validating swap+send state (#24568)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24568?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

ejwessel committed:
Fix issue 24464 (#24570)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24570?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24464

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Signed-off-by: Ethan Wessel <ethan.wessel@consensys.net>

Commit:
docs: add comment for out of scope bug

Commit:
fix: validate on amount error state change

Commit:
feat: submit both transactions on confirm

Commit:
chore: route to tx screen for swap+send

Commit:
feat: add loading state

blocks additional submissions

Commit:
perf: debounce amount history entries

Commit:
fix: ignore casing for address equality

Commit:
chore: use cleanup for state change

Commit:
chore: go to activity tab after submitting

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -2090,6 +2090,10 @@ const debouncedComputeEstimatedGasLimit = debounce((dispatch) => {
 
+const debouncedAddHistoryEntry = debounce((dispatch, payload) => {
+  dispatch(addHistoryEntry(payload));
+}, 100);
```
what's the reasoning behind 100? what's the benefit of debouncing?

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -1665,15 +1712,22 @@ const slice = createSlice({
             bestQuote &&
-              bestQuote.recipient !== draftTransaction.recipient.address,
+              !isEqualCaseInsensitive(
+                bestQuote.recipient.toLowerCase(),
+                draftTransaction.recipient.address.toLowerCase(),
```
```suggestion
                bestQuote.recipient,
                draftTransaction.recipient.address,
```

micaelae commented on ui/components/multichain/pages/send/send.js:
```diff
@@ -192,7 +193,6 @@ export const SendPage = () => {
 
     setIsSubmitting(true);
     await dispatch(signTransaction(history));
```
Does the page fail gracefully if signing a swap + send transaction/approval fails? Example concerns: page getting stuck in loading state, user doesn't get feedback about the failed signing

Maybe `history.push(DEFAULT_ROUTE);` in signTransaction should happen in a finally block?

micaelae commented on ui/components/multichain/pages/send/send.js:
```diff
@@ -183,7 +191,8 @@ export const SendPage = () => {
     event.preventDefault();
 
-    await dispatch(signTransaction());
+    setIsSubmitting(true);
+    await dispatch(signTransaction(history));
```
Does the page fail gracefully if signing a swap + send transaction/approval fails? Example concerns: page getting stuck in loading state, user doesn't get feedback about the failed signing

BZahory replied:
It did not – updated

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -2862,31 +2867,86 @@ export function signTransaction() {
       if (isSwapAndSend) {
-        transactionType = TransactionType.contractInteraction;
-      }
+        const NATIVE_CURRENCY_DECIMALS = 18;
```
Why not use SWAPS_CHAINID_DEFAULT_TOKEN_MAP[chainId].decimals?

BZahory replied:
Fair point – added

Commit:
fix: prevent input overflow-push on firefox

Commit:
fix: hide max button when native send impossible

Commit:
chore: show "confirm" for swap+send

there is no confirmation screen so this adds clarity

Commit:
fix: prevent infinite loading state

Commit:
chore: format modal balances

Commit:
fix: only reset on impossible gas if max

Commit:
refactor: get decimals from object

BZahory committed:
chore: remove redundant toLowerCase()

Co-authored-by: micaelae <100321200+micaelae@users.noreply.github.com>

Commit:
chore: upgrade transaction controller

adds swap+send types

Commit:
chore: add activity log text

Commit:
feat: add swap+send type

treat as swap *or* send transaction when appropriate
add paired icon
add paired title

Commit:
feat: add swap+send handling in tx hook

Commit:
feat: add swap+send activity details

Commit:
chore: add type

Commit:
feat: add tooltip for swap+send transactions

micaelae committed:
[METABRIDGE-843] chore: disable blocked tokens (#24591)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**
Changes
* adds utilities for fetching blockedTokens from the swaps api
* disables a dest token if it's in the blockedTokens list AND if it's
not the same as the src token

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24591?quickstart=1)

## **Related issues**

Fixes: N/A

## **Manual testing steps**

1. Go to send page, chainId=1
2. ETH -> BNB should not be possible
3. BNB -> BNB should be possible
4. BNB -> ETH should not be possible
5. BNB -> USDC should not be possible
6. BNB -> TKN should not be possible

1. Switch to a chain that's not supported by swaps
2. There should be no errors/crashes when attempting to do a simple send
flow
3. Console/Network tab should have no errors regarding the blockedTokens
api call


## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

Blocked dest tokens disabled:


https://github.com/MetaMask/metamask-extension/assets/100321200/cf9afe8f-25e1-4dda-9260-a7c04554b8d3

Disable swap+send when src token is blocked:

<img width="393" alt="Screenshot 2024-05-21 at 4 38 04 PM"
src="https://github.com/MetaMask/metamask-extension/assets/100321200/7bf365b5-4515-4aad-8a12-796cf75cf0d2">





## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

micaelae commented on ui/components/multichain/asset-picker-amount/asset-picker-amount.tsx:
```diff
@@ -61,6 +65,18 @@ export const AssetPickerAmount = ({
+    }
+    onAmountChange('0x0');
+  }, [isNativeSendPossible]);
+
```
Can you add a comment for what this does? Also, why is it needed?

micaelae commented on ui/components/multichain/pages/send/send.js:
```diff
@@ -248,6 +248,9 @@ export const SendPage = () => {
     [dispatch],
   );
 
+  const isSwapAndSend =
+    transactionAsset?.details?.address !== receiveAsset?.details?.address;
+
```
Minor: can be extracted to a util or selector

BZahory replied:
Done

Commit:
feat: use actual destination value

Commit:
refactor: move slippage to util

Commit:
chore: prevent double action call

Commit:
feat: account for gas in best quote selection

Commit:
refactor: create sender selector

Commit:
docs: remove resolved comments

Commit:
docs: resolve final todo and update task list

Commit:
style: appease lint gods: the post credit scene

Commit:
style: fix scss spacing

Commit:
docs: add jsdoc comment

Commit:
refactor: create util

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -3333,6 +3315,19 @@ export function sendAmountIsInError(state) {
+export function getSender(state) {
+  const sendState = state[name];
+
+  const draftTransaction =
+    sendState.draftTransactions[sendState.currentTransactionUUID];
```
can this ever be undefined?

BZahory replied:
Yep  – good catch

micaelae commented on ui/ducks/send/swap-and-send-utils.ts:
```diff
@@ -140,6 +140,9 @@ const QUOTE_VALIDATORS = [
   },
 ];
+
+const SWAP_AND_SEND_SLIPPAGE = '2';
```
just confirming, does swaps have a default slippage and is it the same as this?

BZahory replied:
I couldn't find one. I just found this code in `review-quote.js`:

```
const [slippageErrorKey] = useState(() => {
    const slippage = Number(fetchParams?.slippage);
    if (slippage > 0 && slippage <= 1) {
      return SLIPPAGE_LOW_ERROR;
    } else if (slippage >= 5 && slippage <= MAX_ALLOWED_SLIPPAGE) {
      return SLIPPAGE_HIGH_ERROR;
    }
    return '';
  });
```

This indicates that the range for swaps quotes is 1-5 (both exclusive). I'd imagine we'd want to be on the lower end given the typical use cases for a send, so 2 seems like a clean number to choose.

Commit:
chore: prevent direct quote mutation

Commit:
chore: add undefined guards for `calcTokenAmount`

legobeat reviewed (dismissed):
Could you update PR description to contain a summary of the intended change(s) done in this PR, as well as a briefer one-line summary in the title? Seems like this is targeting some other open PR?

[MetaMask Guide to Pull Requests](https://github.com/MetaMask/contributor-docs/blob/main/docs/pull-requests.md)

[Also, unit tests are currently failing](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/82055/workflows/73c8f2c3-ff6b-42db-9d4f-5ce283667dab/jobs/2917780)

Commit:
fix: clear send tx on swap+send submit

Commit:
chore: add optional chain

Commit:
perf: do not focus input if an input is already focused

Commit:
chore: remove unused message

georgewrmarshall commented on ui/components/component-library/button-icon/button-icon.scss:
```diff
@@ -29,6 +29,10 @@
   }
 
   // ButtonIcon Sizes
+  &--size-xs {
+    --button-icon-size: 20px;
+  }
```
Should this be removed? Started a discussion in slack regarding changes in https://github.com/MetaMask/metamask-extension/pull/24150

BZahory replied:
Will address both comments in a follow-up PR in the merge train

georgewrmarshall commented on ui/components/component-library/button-link/button-link.scss:
```diff
@@ -58,7 +58,7 @@
   }
 
-  &--disable-underline:hover:not(&--disabled){
+  &--disable-underline:hover:not(&--disabled) {
```
See comment above

micaelae commented on ui/ducks/send/send.js:
```diff
@@ -2804,7 +2806,8 @@ export function signTransaction(history) {
 
-    if (stage === SEND_STAGES.EDIT) {
+    // you can only edit a basic send transaction
+    if (stage === SEND_STAGES.EDIT && !isSwapAndSend) {
```
Does this mean we can remove the prev swap send state from redux as well?

BZahory replied:
In theory but I’d rather keep it for now since we may be reutilizing it p soon. A bit of a boat anchor that we can clear pending future discussion on confirmations

Commit:
fix: prevent textarea focus steal
